File: 094_ChildrenSumProperty.cpp
/*
Problem: Children Sum Property
Description: Check if sum of children equals parent value for all nodes.

Approach: Recursively check condition at each node.
*/

#include <bits/stdc++.h>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left, *right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

bool isChildrenSum(TreeNode* root) {
    if (!root) return true;
    
    if (!root->left && !root->right) return true;
    
    int sum = 0;
    if (root->left) sum += root->left->val;
    if (root->right) sum += root->right->val;
    
    return (sum == root->val) && 
           isChildrenSum(root->left) && 
           isChildrenSum(root->right);
}

int main() {
    TreeNode* root = new TreeNode(10);
    root->left = new TreeNode(8);
    root->right = new TreeNode(2);
    root->left->left = new TreeNode(3);
    root->left->right = new TreeNode(5);
    
    cout << isChildrenSum(root) << endl;  // Output: 1 (true)
    
    return 0;
}

/*
Time Complexity: O(n)
Space Complexity: O(h)
*/

File: 094_ChildrenSumProperty.py
"""
Problem: Children Sum Property
Description: Check if sum of children equals parent value for all nodes.

Approach: Recursively check condition at each node.
"""

from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def isChildrenSum(root: Optional[TreeNode]) -> bool:
    if not root:
        return True
    
    if not root.left and not root.right:
        return True
    
    total = 0
    if root.left:
        total += root.left.val
    if root.right:
        total += root.right.val
    
    return (total == root.val and 
            isChildrenSum(root.left) and 
            isChildrenSum(root.right))

def main():
    root = TreeNode(10)
    root.left = TreeNode(8)
    root.right = TreeNode(2)
    root.left.left = TreeNode(3)
    root.left.right = TreeNode(5)
    
    print(isChildrenSum(root))  # Output: True

if __name__ == "__main__":
    main()

"""
Time Complexity: O(n)
Space Complexity: O(h)
"""
